src/handle/tactic.ml

```diff
@@ -97,19 +97,23 @@ let tac_solve : popt -> proof_state -> proof_state = fun pos ps ->
 
 (** [tac_refine pos ps gt gs p t] refines the typing goal [gt] with [t]. [p]
    is the set of metavariables created by the scoping of [t]. *)
-let tac_refine :
+let tac_refine : ?check:bool ->
       popt -> proof_state -> goal_typ -> goal list -> problem -> term
       -> proof_state =
-  fun pos ps gt gs p t ->
+  fun ?(check=true) pos ps gt gs p t ->
   if Logger.log_enabled () then
     log_tact "@[tac_refine@ %a@]" term t;
   let c = Env.to_ctxt gt.goal_hyps in
   if LibMeta.occurs gt.goal_meta c t then fatal pos "Circular refinement.";
   (* Check that [t] has the required type. *)
-  match Infer.check_noexn p c t gt.goal_type with
-  | None ->
-      fatal pos "%a@ does not have type@ %a."  term t  term gt.goal_type
-  | Some t ->
+  let t =
+    if check then
+      match Infer.check_noexn p c t gt.goal_type with
```

Comment:
Shouldn't `Query.check` be used here?

Reply:
I don't think so because we do not want to call solve at this stage yet. Solve is called after the refinement on all constraints generated by this check + the previously existing constraints, while Query.check tries to solve only the constraints of p which does not include the previously existing constraints (since we need to know what are the new ones). What do you think?

Reply:
But isn't it the case that the problem `p` already contains constraints generated previously? How come `p` does not include previous constraints? There is only one `new_problem ()` in this section, isn't it?

Reply:
p is given as argument to tac_refine. It is used to record new metas/goals and new constraints when applying a tactic. See for instance tac_induction. Hence also the call to add_goals_of_problem.

Reply:
Sure, but the  `p` given to `tac_refine` contains everything collected so far, so refine must be able to say whether the unification problem is satisfiable or not. Or are there  cases where a unification problem is provably not satisfiable, but we don't want `tac_refine` to fail?

Reply:
Exactly: unsolvable constraints are added as new goals.

Reply:
Ah I'm afraid I was confused and thought of
https://github.com/Deducteam/lambdapi/blob/f89b2c32e397f0fc3860b48c6d4ad99b18c47fd7/src/handle/command.ml#L419

Reply:
So it seems clear for this line and PR, no problem. For the one I just pasted, it may be discussed somewhere else.

Reply:
In command.ml, we need to do something on t to replace place holders by metas.
But that's right that we should fail if check returns None here.

Reply:
This is fixed in https://github.com/Deducteam/lambdapi/pull/849/commits/5f3fb142395150308aace5b64aeb5ad95606c2a9.
But we may wonder on the contrary why we could not replace the calls to Query.check_sort and Query.infer by calls to Infer.check_sort_noexn and Infer.check_noexn so as to be more general since the generated unsolved constraints could perhaps be solved by tactic applications. What do you think?

Reply:
Well, currently, this would make no difference because we have no tactic other than solve for unification goals. But this could happen later.

Reply:
Done in https://github.com/Deducteam/lambdapi/pull/849/commits/1050abd43ba0737a15ea096bd8cf9551fd0c971a. Your comments are welcome.